File: exp-02/q2-binary-search-recursion/solution.go
package main

import "fmt"

func binarySearch(arr []int, low, high, key int) int {
	if low > high {
		return -1
	}

	mid := low + (high-low)/2

	if arr[mid] == key {
		return mid
	} else if arr[mid] > key {
		return binarySearch(arr, low, mid-1, key)
	} else {
		return binarySearch(arr, mid+1, high, key)
	}
}

func main() {
	var n, key int
	fmt.Scan(&n)

	arr := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&arr[i])
	}

	fmt.Scan(&key)

	index := binarySearch(arr, 0, n-1, key)

	if index == -1 {
		fmt.Println("Element not found")
	} else {
		fmt.Printf("Element found at index %d\n", index)
	}
}

File: exp-02/q2-binary-search-recursion/Solution.java
import java.util.Scanner;

public class Solution {
    // Recursive binary search
    static int binarySearch(int[] arr, int low, int high, int key) {
        if (low > high) {
            return -1;
        }

        int mid = low + (high - low) / 2;

        if (arr[mid] == key) {
            return mid;
        } else if (arr[mid] > key) {
            return binarySearch(arr, low, mid - 1, key);
        } else {
            return binarySearch(arr, mid + 1, high, key);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] arr = new int[n];

        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        int key = sc.nextInt();

        int index = binarySearch(arr, 0, n - 1, key);

        if (index == -1) {
            System.out.println("Element not found");
        } else {
            System.out.println("Element found at index " + index);
        }
    }
}